bgpd/bgp_fsm.c

```diff
@@ -1249,8 +1250,12 @@ static int bgp_connect_check(struct thread *thread)
 
 	/* Check file descriptor. */
 	slen = sizeof(status);
+	if (peer && peer->bgp)
+		vrf_switch_to_netns(peer->bgp->vrf);
```

Comment:
This construct of having the protocol daemons having to switch back and forth for every socket call is another construct that will kill any hope of maintainability for me.  If getsockopt, or really any OS call, needs to be called in the the `correct` namespace than make a wrapper function that is called that we pass the vrf in, because why the hell should bgp care about this?

Reply:
you understand that if we choose a wrapper, then you will have following consequences:
-1 there may be more changes.  ( because many socket operations are called, whereas before the switch/ is called in wise way)
-2 for instance, the vrf pointer ( or vrf_id) will have to be passed for every socket operation.
-3 on the 1 case, because the changes will move to the unitary socket operations, then ther may be several calls that will be done. ( you may lose the errno value also).

Reply:
1. I would rather have more changes in one patch than have to manually turn every socket related system call into 5 lines instead of 1, forever.
2. No comment
3. This can be designed around.

That said, I'm not even sure a wrapper is the best way to solve this. I feel like the fact that any of this is necessary indicates some underlying design issues. Why can't we do this patch behind the existing VRF implementation, totally transparent to daemons? Why do they have to know about namespaces at all? They know about the concept of VRFs, and if your aim is to emulate VRFs with namespaces, I don't think its the daemon's responsibility to learn about namespaces as such -- I think it's the responsibility of the layer that exposes the concept of a VRF to the daemon to learn about a new way of implementing VRFs behind the scenes. This way the daemons don't need to know about namespaces, the same way they currently don't know about VRF-lite -- they just know about VRF in its Platonic form.

In terms of this immediate pull request, my conviction is that forcing the programmer to put these incantations around every socket-related system call is unacceptable...

Reply:
Indeed that would be unacceptable. The good news is that most of the changes in bgpd are unnecessary and can be removed. We only need to do the `setns()` dance when creating sockets with the `socket()` syscall. For all other syscalls (e.g. `write()`, `setsockopt()`), the Linux kernel is capable of identifying the netns of the socket on its own, because it retains this information at the moment the socket is created.

We also need to use `setns()` around the ioctl calls (e.g. `ioctl(sd, SIOCETHTOOL, ...)` otherwise a few things won't work as expected.